drivers: misc: devmux: add a device multiplexer pseudo device

cfriedt opened the pull request:
The Device Multiplexer (`devmux`) is a pseudo-device that operates as a device switch, allowing the user to select which device is used under the hood. The `devmux` device looks and feels exactly like the selected device.

```
            +----------+                            +----------+
            |  devmux  |                            |  devmux  |
            |          |                            |          |
 dev0       |          |                 dev0       |          |
 +---------->   \      |                 +---------->          |
            |    \     |                            |          |
 dev1       |     \    |       dev0      dev1       |          |       dev2
 +---------->      O   +---------->      +---------->      O   +---------->
            |          |                            |     /    |
 dev2       |          |                 dev2       |    /     |
 +---------->          |                 +---------->   /      |
            |          |                            |          |
            |          |                            |          |
            |          |                            |          |
            +-----^----+                            +-----^----+
                  |                                       |
   select == 0    |                       select == 2     |
   +--------------+                       +---------------+

```

Devmux devices are declared and configured via Devicetree. For example,
```dts
/ {
    chosen {
        zephyr,console = &devmux;
        zephyr,shell_uart = &devmux;
    };

    euart0: uart_emul0 {
        compatible = "zephyr,uart-emul";
        current-speed = <0>;
        status = "okay";
    };

    euart1: uart_emul1 {
        compatible = "zephyr,uart-emul";
        current-speed = <0>;
        status = "okay";
    };

    devmux: devmux {
        compatible = "zephyr,devmux";
        devices = <&uart0 &euart0 &euart1>;
        status = "okay";
        zephyr,mutable;
    };
};
```

It is experimental, but its current use is in system diagnostics and remediation. In particular, `devmux` is particularly suitable when a fallback is required for a singleton device in Zephyr, such as the console or shell uart.

Take for example, the scenario where the system console and log subsystem both have the uart backend enabled. Assume we have chosen the backing uart to be an abstraction of some other high-bandwidth bus - such as a PCIe BAR, a UDP socket, or even even just memory.

If the "service" (for lack of a better term) that backs this abstract "uart" experiences an error and further system diagnostics are needed for remediation, it is incredibly helpful to be able to switch the system console, shell uart, or whatever to another uart in order to have continued use of the shell, examine system status, to view system logs, etc.

Test Output:
```shell
west build -p auto -b qemu_riscv64 -t run tests/drivers/console_switching/
...
*** Booting Zephyr OS build zephyr-v3.4.0-3988-gaaefb2d764ea ***
Running TESTSUITE console_switching
===================================================================
START - test_read
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
 PASS - test_read in 0.005 seconds
===================================================================
START - test_write
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
 PASS - test_write in 0.003 seconds
===================================================================
TESTSUITE console_switching succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [console_switching]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.008 seconds
 - PASS - [console_switching.test_read] duration = 0.005 seconds
 - PASS - [console_switching.test_write] duration = 0.003 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Commit:
device: support for mutable devices

Add support for mutable devices. Mutable devices are those which
can be modified after declaration, in-place, in kernel mode.

In order for a device to be mutable, the following must be true

  * `CONFIG_DEVICE_MUTABLE` must be y-selected
  * the Devicetree bindings for the device must include
    `mutable.yaml`
  * the Devicetree node must include the `zephyr,mutable` property

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
drivers: misc: devmux: a device multiplexer pseudo-device

The Device Multiplexer (devmux) is a pseudo-device that can
be used to select between multiple included sub-devices.

It is experimental, but its current use is in system
remediation. Take for example, the scenario where the
system console and log subsystem both have the uart backend
enabled. The case may arise, where the chosen backing uart
could be an abstraction of another very high-bandwidth bus
- such as a PCIe BAR, a UDP socket, or even even just memory.

If the "service" (for lack of a better term) that backs this
abstract "uart" experiences an error, it is of critical
importance to be able to switch the system console, uart log
backend, or whatever to another uart (semi-transparently) in
order to bring up a shell, continue to view system logs, or
even just support user console I/O.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: drivers: console: add console_switching test

This testsuite exercises a number of things. Namely

  * `mutable` devices (i.e. those backed in SRAM)
  * `uart_emul` support for interrupt mode (receive only)
  * `devmux` capabilities of multiplexing several uarts
  * switching the system console between several uart backends

Testing Done:
```
west build -p auto -b qemu_riscv64 -t run \
  tests/drivers/console_switching/
...
*** Booting Zephyr OS build zephyr-v3.4.0-3988-gaaefb2d764ea ***
Running TESTSUITE console_switching
================================================================
START - test_read
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
 PASS - test_read in 0.005 seconds
================================================================
START - test_write
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
 PASS - test_write in 0.003 seconds
================================================================
TESTSUITE console_switching succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [console_switching]: pass = 2, fail = 0,...
 - PASS - [console_switching.test_read] duration = 0.005 seconds
 - PASS - [console_switching.test_write] duration = 0.003 seconds

------ TESTSUITE SUMMARY END ------

===============================================================
PROJECT EXECUTION SUCCESSFUL
```

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* add Meta copyright to `mutable.h`

cfriedt commented:
* fill-in userspace support

cfriedt commented:
* added a diagram to doxygen via https://asciiflow.com

cfriedt commented:
* added `testcase.yaml`

fabiobaltieri commented on include/zephyr/drivers/misc/devmux/devmux.h:
```diff
@@ -0,0 +1,88 @@
+ *            |          |                            |     /    |
+ * dev2       |          |                 dev2       |    /     |
+ * +---------->          |                 +---------->   /      |
+ *            |          |                            |          |
```
drop a line? 3 up, 4 down... the asymmetry is killing me

cfriedt replied:
Not sure what you mean

fabiobaltieri replied:
ah that's fine, the `devmux` label is centered so I guess you have to draw around that... ascii art is hard!

cfriedt replied:
Fixed!

ycsin commented on drivers/serial/uart_emul.c:
```diff
@@ -127,6 +230,15 @@ uint32_t uart_emul_put_rx_data(const struct device *dev, uint8_t *data, size_t s
+
+	if (count > 0 && drv_data->irq_en) {
+
+		drv_data->irq_en = true;
```
`drv_data->irq_en` should be already `true` if the conditionals above is `true`

cfriedt replied:
Fixed!

cfriedt commented:
* rebase  after fix for #62999 was merged
* remove redundant line in `uart_emul.c`

gmarull commented on drivers/serial/uart_emul.c:
```diff
@@ -99,15 +112,105 @@ int uart_emul_config_get(const struct device *dev, struct uart_config *cfg)
-};
+	.err_check = uart_emul_err_check,
+
+	COND_CODE_1(
+		IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN),
```
`IF_ENABLED` would simplify this

cfriedt replied:
moved to and fixed in https://github.com/zephyrproject-rtos/zephyr/pull/63416

gmarull commented on dts/bindings/base/mutable.yaml:
```diff
@@ -0,0 +1,13 @@
+# Properties for Mutable devices (PM)
+
+properties:
+  mutable:
+    type: boolean
+    description: |
+      True iff the device structure may be mutated.
+
+      Inherit this binding for devices that are runtime-modifiable, in-place.
+      This places the device structure into SRAM rather than Flash.
```
this is not hardware description, so at least it should be prefixed with `zephyr,`

cfriedt replied:
I don't necessarily see the other draft PR as necessarily blocking this, but I would be happy to minimize whatever might be necessary to fix.

cfriedt replied:
Fixed!

gmarull commented on include/zephyr/device/mutable.h:
```diff
@@ -0,0 +1,166 @@
+	static const Z_DECL_ALIGN(struct init_entry) __used __noasan           \
+		Z_INIT_ENTRY_SECTION(level, prio,                              \
+				     Z_DEVICE_INIT_SUB_PRIO(node_id))          \
+		Z_INIT_ENTRY_NAME(DEVICE_NAME_GET(dev_id)) = {                 \
+			.init_fn = {.dev_rw = (init_fn_)},                        \
+			.dev_rw = (struct device *)&DEVICE_NAME_GET(dev_id),                       \
+	}
+
+/**
+ * @brief Define a mutable @ref device and all other required objects.
+ *
+ * This macro is similar to @ref Z_DEVICE_DEFINE but is used for defining
+ * mutable devices.
+ *
+ * @param node_id Devicetree node id for the device (DT_INVALID_NODE if a
+ * software device).
+ * @param dev_id Device identifier (used to name the defined @ref device).
+ * @param name Name of the device.
+ * @param init_fn Device init function.
+ * @param pm Reference to @ref pm_device associated with the device.
+ * (optional).
+ * @param data Reference to device data.
+ * @param config Reference to device config.
+ * @param level Initialization level.
+ * @param prio Initialization priority.
+ * @param api Reference to device API.
+ * @param state Reference to device state.
+ * @param ... Optional dependencies, manually specified.
+ */
+#define Z_DEVICE_DEFINE_MUT(node_id, dev_id, name, init_fn, pm, data, config,      \
+			level, prio, api, state, ...)                          \
+	Z_DEVICE_NAME_CHECK(name);                                             \
+                                                                               \
+	IF_ENABLED(CONFIG_DEVICE_DEPS,                                         \
+		   (Z_DEVICE_DEPS_DEFINE_MUT(node_id, dev_id, __VA_ARGS__);))      \
+                                                                               \
+	Z_DEVICE_BASE_DEFINE_MUT(node_id, dev_id, name, pm, data, config, level,   \
+			     prio, api, state, Z_DEVICE_DEPS_NAME(dev_id));    \
+                                                                               \
+	Z_DEVICE_INIT_ENTRY_DEFINE_MUT(node_id, dev_id, init_fn, level, prio)
```
I'm a bit skeptical about all this complexity. Why is this needed? Isn't the dt flag enough?

cfriedt replied:
IIRC, this was the minimum required to support 
`DEVICE_DT_INST_DEFINE_MUT()` with a separately named iterable section (`devices_mutable` instead of `devices`). I.e. there was a warning or other error when using the same name, but I can double check.

Also, it kind of makes sense to have a separate iterable sections because the memory for `devices` is in `rodata` (rom) while the memory for `devices_mutable` is in `data` (ram) and they should really be independently contiguous.

gmarull replied:
So are devices forced to use `DEVICE_DT_DEFINE_MUT` to be mutable? I don't think this will scale well, specially because there's an ongoing trend to have class-specific device definition macros (e.g. can, i2c, etc.), and the situation there is already pretty bad in terms of complexity. Isn't a `zephyr,mutable` property enough to decide wether to place the device in a ram section instead of rom? Also note that there's code depending on all devices being on a single list (PM_DEVICE), not sure if this can affect it.

cfriedt replied:
Technically speaking we could put logic inside the existing macros. I just wanted to avoid being overly intrusive.

I can respin if that is easier, and then, yeah, it would still use the existing interface.

cfriedt replied:
Fixed!

gmarull commented on include/zephyr/init.h:
```diff
@@ -96,7 +107,12 @@ struct init_entry {
+#ifdef CONFIG_DEVICE_MUTABLE
+		struct device *dev_rw;
+#endif
+	};
```
I think this should wait for all the changes scheduled for the init changes https://github.com/zephyrproject-rtos/zephyr/pull/61986, otherwise we have yet another thing on the table to fix.

cfriedt replied:
I don't see the other draft PR as necessarily blocking this, but I would be happy to minimize whatever might be necessary to fix. E.g. originally, I added the macros from mutable.h directly into device.h because I thought it might be nice to be slightly separate. I can move them back if it's simpler to be in the same file.

gmarull replied:
after looking at it again, it seems that it should be easy to port, since the proposal removes devices from init.h.

gmarull commented on kernel/Kconfig:
```diff
@@ -1083,6 +1083,13 @@ config DEVICE_DEPS_DYNAMIC
+	bool "Mutable devices [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  Support mutable devices. Mutable devices are instantiated in SRAM
+	  instead of Flash and are runtime modifiable in kernel mode.
```
I think mutable devices should be documented clearly to avoid future misusages. This is a bit too generic IMHO.

cfriedt replied:
> I think mutable devices should be documented clearly to avoid future misusages.

Clearly documented where / how? Please feel free to suggest a specific preference.

> This is a bit too generic IMHO.

Hehe.. well, it's definitely possible to specialize things for more specialized use cases, but I think that might be getting into "premature specialization".

Maybe a DT property could be added allowing one to specify the initial selection. But at the end of the day, it's a multiplexer (switch) - there really isn't much more to the basic requirement.

Also, we have devicetree to aid in specialization through inheritance. So if specialization needs to happen for a particilar subclass, it might be better to add e.g. new properties as necessary, rather than forcing them in the base case.

kartben replied:
@cfriedt I'd really like to see a code sample (+README)  actually. For example could this be used to transparently expose a modem's AT shell as zephyr,shell-uart so that one can tinker with AT commands directly and then, with the push of a button, one can go back to the Zephyr shell? Or am I completely off?

kartben replied:
And if I'm not off then I guess the sample should also showcase the simpler version where one only changes "selected" in the devicetree. The button thing is just to show the dynamic aspect.

cfriedt replied:
Oh weird - my comment didn't seem to submit properly from the GH app. 

> one can go back to the Zephyr shell? Or am I completely off?

Yeah, that's basically the gist of it, although there are many applications. 

>  I'd really like to see a code sample (+README) actually. 

There is the existing test obviously that shows what's necessary to switch between multiplexed uart devices in software. Extrapolating that out to a sample app is pretty trivial.

>  the simpler version where one only changes "selected" in the devicetree.

Changing `selected` in Devicetree just allows for explicit specification of the initial selection (which is otherwise set to position zero). The devmux device is still fully functional and isn't necessarily any simpler.

I guess a sample app would help to clarify those things, so I will get to it 😁

kartben replied:
Did you post in the wrong comment thread? If yes could you move the comment above to the other PR for reference? FWIW my understanding is that we're not even adding new code samples during feature freeze so I am not sure I would agree with your statement.

cfriedt replied:
I thought you had asked for a sample. At first, I thought sure, but as you mentioned we are past feature freeze. I'd prefer not to add a sample here in this case (mainly because this change is really just being used for recovering unresponsive Zephyr FW).

kartben replied:
> I thought you had asked for a sample. At first, I thought sure, but as you mentioned we are past feature freeze. I'd prefer not to add a sample here in this case (mainly because this change is really just being used for recovering unresponsive Zephyr FW).

Oh yes I'm totally asking for a sample, sorry if this wasn't clear! But I guess the misunderstanding is that I, and I believe everyone else in this thread, is just discussing in the context of 3.6, to make sure this is ready to go as soon as the doors open again.

I don't see how this can make it for 3.5, unless this was discussed in today's release meeting which I missed? But this would also require a TSC vote IIUC.

gmarull commented:
Another comment: `device_is_ready()` should likely return true only if the muxed device is actually ready. Does it make sense?

cfriedt commented:
Going to separate the `uart_emul` changes to narrow the scope of this PR, will address review comments, and then respin.

cfriedt commented:
> Another comment: `device_is_ready()` should likely return true only if the muxed device is actually ready. Does it make sense?

That's always been the case, I think. There is a loop that checks ready status for all of the referenced devices in `devmux_init()`.

cfriedt commented:
* added DT property `selected` (default 0), to specify the initial selection of the devmux

cfriedt commented:
retested
```bash
west build -p auto -b qemu_riscv64 -t run tests/drivers/console_switching
...
*** Booting Zephyr OS build v3.5.0-rc1-38-g3ced117c6ea2 ***
Running TESTSUITE console_switching
===================================================================
START - test_read
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
 PASS - test_read in 0.005 seconds
===================================================================
START - test_write
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
 PASS - test_write in 0.003 seconds
===================================================================
TESTSUITE console_switching succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [console_switching]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.008 seconds
 - PASS - [console_switching.test_read] duration = 0.005 seconds
 - PASS - [console_switching.test_write] duration = 0.003 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

cfriedt commented:
* added the `selected` DT property to specify initial selection, defaulting to 0
* added `devmux1` node to test overlay to test `selected` property

cfriedt commented:
* whitespace corrections

faxe1008 commented on dts/bindings/base/mutable.yaml:
```diff
@@ -0,0 +1,13 @@
+  zephyr,mutable:
+    type: boolean
+    description: |
+      True iff the device structure may be mutated.
```
Small typo

cfriedt replied:
> Small typo

"Iff" is a mathematical term meaning "if and only if"

faxe1008 replied:
okay, did not know that :^)

faxe1008 commented on drivers/misc/devmux/Kconfig:
```diff
@@ -0,0 +1,23 @@
+	  software to select the data, config, and api from a number of linked
+	  devices.
+
+if DEVMUX
```
Nit: make it a `depends on` instead.

cfriedt replied:
I usually see `depends on` when the dependency is from a separate `Kconfig` file. It's more common to use `if FOO` in the same file where `FOO` is defined, because it saves space when there are (or it is anticipated that there will be) multiple dependencies in the same file.

fabiobaltieri commented on drivers/misc/devmux/devmux.c:
```diff
@@ -0,0 +1,177 @@
+	devmux_configs[inst] = config;
+	devmux_datas[inst] = data;
+
+	for (size_t i = 0; i < config->n_devs; ++i) {
+		if (!device_is_ready(config->devs[i])) {
+			return -ENODEV;
+		}
+	}
```
Hey would it make sense to not fail the devmux init if some of the muxed devices are not ready and instead fail the select? Was thinking about a use case where some of the devices may not be available (for example NACKing an i2c address or failing an id register check), and then you could use this for finding alternate devices and have some devices not ready as a known good condition.

cfriedt replied:
Hmm... what about the initial selection? I guess it would still make sense to fail on init with the initial selection, right?

Otherwise, that works for me! What do you think, @gmarull?

fabiobaltieri replied:
Oh I see the problem. Yeah not sure what to do about the initial selection. Problem is that if you fail the init then the devmux device is flagged as not ready and you can't use it at all. Maybe it make sense to always initially select the first ready sub-device? (if any, if none are ready just fail the mux device as well)

cfriedt replied:
With initial selection, it actually narrows the scope of initialization checks, which I think is a good thing. Technically speaking, only the initially selected device needs to be ready by the time the devmux is initialized.

But to your point, it absolutely does make sense to allow possibly tardy devices on i2c, spi, or other buses to have their readiness checked only before selection.

I think that's a quite reasonable compromise. At some point in the future, maybe it would make sense to even allow for deferred initial selection (maybe via DT property?). I'm in no rush to add that though, as my use case primarily involves the console and shell.

cfriedt commented:
* reworked init to only error if the initially selected device is not ready
* reworked select_set to verify that the selected device is ready
* added -ENODEV retval to select_set when device is not ready

cfriedt commented:
Rebased after #63416 was merged - I'll see if I can come up with a straight-forward dual-uart sample on a common board as well.

cfriedt commented:
I'd like to propose getting this in for 3.5.0, so will tag it for TSC

kartben commented:
> I'd like to propose getting this in for 3.5.0, so will tag it for TSC

You'll probably want to bring it up on Discord and/or mailing list since there's no TSC meeting this week and this probably needs to be decided upon before RC2?

gmarull commented on include/zephyr/linker/common-ram.ld:
```diff
@@ -141,6 +141,10 @@
 	ITERABLE_SECTION_RAM(zbus_channel_observation_mask, 1)
 #endif /* CONFIG_ZBUS */
 
+#if defined(CONFIG_DEVICE_MUTABLE)
+	ITERABLE_SECTION_RAM(device_mutable, 4)
+#endif
```
nit: you need to update linker generator cmake files as well.

cfriedt replied:
Fixed!

gmarull commented on include/zephyr/sys/iterable_sections.h:
```diff
@@ -234,6 +234,16 @@ extern "C" {
+ * The name can be used to customize how iterable section entries are sorted.
+ * @see STRUCT_SECTION_ITERABLE_NAMED()
+ */
+#define STRUCT_SECTION_ITERABLE_NAMED_ALTERNATE(struct_type, secname, name, varname) \
+	TYPE_SECTION_ITERABLE(struct struct_type, varname, secname, name)
```
maybe update iterable sections test?

cfriedt replied:
Fixed!

gmarull commented on drivers/misc/devmux/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+
+zephyr_library()
+
+zephyr_syscall_header(
+  ${ZEPHYR_BASE}/include/zephyr/drivers/misc/devmux/devmux.h
+)
```
~is this strictly required? (sorry, was not aware of this)~

edit: I guess it's needed because there's no `__subsystem` tag.

cfriedt replied:
It is needed otherwise the system calls aren't generated. I don't know the exact details of the machinery but we added these when @dcpleung made some changes a while back to only generate the minimal number of syscalls

gmarull commented on dts/bindings/misc/zephyr,devmux.yaml:
```diff
@@ -0,0 +1,20 @@
+      Devices to be multiplexed.
+  selected:
+    type: int
+    default: 0
+    description: |
+      Initial multiplexer selection.
```
per guidelines, please explain/justify default value.

cfriedt replied:
Fixed!

gmarull reviewed (commented):
Overall lgtm, just a few nitpicks. I think the current revision is much simpler than before. Thanks!

cfriedt commented:
Copy-pasting my comment from #63416

> Right but this definitely sounds to me like it's more than "stabilization and cosmetic changes, bug and doc fixes".

This was used to diagnose and fix an infinite loop in an isr (#63000) that was a problem in just about every Zephyr device in existence. #63000 was split from #62998 to make code easier to review.

#62998 itself is used to improve diagnostics and system remediation on devices that serve billions of people every day.

> In any case, this should have probably followed the minimum review period of 2 days.

Technically speaking, this code has already been already under review for 2 weeks. This was split from #62998 to make code easier to review.

I appreciate your concern, but these were far from a cosmetic changes.

Commit:
tests: misc: iterable_sections: add named alternate tests

Add tests for the newly added
STRUCT_SECTION_ITERABLE_NAMED_ALTERNATE(). This type of
section should be iterated with
STRUCT_SECTION_FOREACH_ALTERNATE().

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
iterable_sections: define iterable_named_alternate variant

Add STRUCT_SECTION_ITERABLE_NAMED_ALTERNATE() for structures
ordered by name in a custom section.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* addressed review comments
* separated iterable sections change to a separate commit
* added iterable sections test in a separate commit
* rebased
* retested

gmarull commented on drivers/misc/devmux/devmux.c:
```diff
@@ -0,0 +1,180 @@
+
+DT_INST_FOREACH_STATUS_OKAY(DEVMUX_DEFINE)
+
+#define DEVMUX_DEVICE_GET(_n) DEVICE_DT_GET(DT_DRV_INST(_n)),
+static const struct device *devmux_devices[] = {DT_INST_FOREACH_STATUS_OKAY(DEVMUX_DEVICE_GET)};
```
`static const struct device *devmux_devices[] = {DT_INST_FOREACH_STATUS_OKAY(DEVICE_DT_INST_GET)};`

cfriedt replied:
Good call

cfriedt replied:
Hmm... that one actually results in a compile error, but works with slight modification.

gmarull commented on drivers/misc/devmux/devmux.c:
```diff
@@ -0,0 +1,180 @@
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(_n, devmux_init, NULL, &devmux_data_##_n,                            \
+			      (void *)&devmux_config_##_n, PRE_KERNEL_1,                           \
```
`(void *)` redundant

cfriedt replied:
Fixed!

gmarull commented on drivers/misc/devmux/devmux.c:
```diff
@@ -0,0 +1,180 @@
+	DEVICE_DT_GET(DT_PHANDLE_BY_IDX(node_id, prop, idx))
+
+#define DEVMUX_PHANDLE_DEVICES(_n)                                                                 \
+	DT_FOREACH_PROP_ELEM_SEP(DT_DRV_INST(_n), devices, DEVMUX_PHANDLE_TO_DEVICE, (,))
```
`DT_INST_FOREACH_PROP_ELEM_SEP(_n, ...)`

cfriedt replied:
Fixed!

cfriedt commented on drivers/misc/devmux/devmux.c:
```diff
@@ -0,0 +1,180 @@
+static const struct devmux_config *devmux_configs[N];
+static struct devmux_data *devmux_datas[N];
+
+static bool devmux_device_is_valid(const struct device *dev)
```
It could be done, but it's not something super critical that we tend to require

cfriedt replied:
I'll give it a shot tho - doesn't hurt necessarily.

cfriedt replied:
Fixed!

cfriedt commented:
* addressed review comments

fabiobaltieri reviewed (approved):
LGTM for the ASCII art part.

galak commented on dts/bindings/base/mutable.yaml:
```diff
@@ -0,0 +1,13 @@
+# Properties for Mutable devices
+
+properties:
+  zephyr,mutable:
```
What's the point of this property.  If `zephyr,devmux` is the only thing that sets it why is this needed?

cfriedt replied:
Technically, any device could set this property but it's really only useful for devices that need to have a runtime modifiable interface for whatever reason.

There are many possible use cases. Some identified in the earlier PR comments, so the utility shouldn't be limited to `zephyr,devmux`.

galak replied:
Does it make sense to add it later?

Is this more descriptive of what it's meaning:

`zephyr,device-api-mutuable`?

cfriedt replied:
I'm ok with the name change, but the property is definitely needed for devmux to work correctly / not fault.

gmarull replied:
I think `zephyr,mutable` is just fine, short and clear.

cfriedt replied:
@galak, @gmarull - I'm ok with naming things either way, but my preference is the short + sweet version.

gmarull replied:
to me PR is ok as is

cfriedt replied:
@galak - I can use your suggestion of `zephyr,device-api-mutuable`. @gmarull - does it matter to you? Would likely need to rebase anyway and get approvals again.

galak reviewed (dismissed):
Why do we need zephyr,mutable property?

carlescufi commented:
Architecture WG:

- @nashif asks what the use of this is. @cfriedt responds that a good example is the console, where you access the console via a devmux device and then if it becomes unresponsive you can switch to a second one via a call to `devmux_select_set()`
- @henrikbrixandersen asks if the new `zephyr,mutable` property adds overhead to the Devicetree code. @cfriedt responds that it doesn't think so
- @henrikbrixandersen also ponders whether we are papering over the lack of flexibility of the console subsystem with this, whether we should instead address those at the root (i.e. console)

nashif commented:
> @[cfriedt](https://github.com/cfriedt) cfriedt self-assigned this [on Sep 23](https://github.com/zephyrproject-rtos/zephyr/pull/62998#event-10453452476)

Please do not assign yourself on new APIs and functionality, let the bot do that or assign someone else.

cfriedt commented:
Statically allocated device nodes for (some maximum number of) hot-pluggable devices is another interesting aspect of the mutable property. E.g. something like USB device enumeration for a USB host controller.

carlescufi commented:
Merging since original assignee has not reviewed in a month.

marc-hb commented on include/zephyr/device.h:
```diff
@@ -961,15 +966,15 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
+		level, prio, Z_DEVICE_INIT_SUB_PRIO(node_id))                                      \
+		Z_INIT_ENTRY_NAME(DEVICE_NAME_GET(dev_id)) = {                                     \
+			.init_fn = {COND_CODE_1(Z_DEVICE_IS_MUTABLE(node_id), (.dev_rw), (.dev)) = \
+					    (init_fn_)},                                           \
```
This initialization syntax seems to break RG-2017.8-linux/XtensaTools/bin/xt-xcc  version 12.0.8, see below.

That toolchain is admittedly old (based on GCC 4 or 5) but it's a pity that an "experimental" feature breaks the build of every .c file even when none of them uses that new feature... is there any C99 compliant way to do the same?

```
13:03:03 zephyr/subsys/logging/log_core.c:944: error: unknown field ‘dev’ specified in initializer
13:03:03 zephyr/subsys/logging/log_core.c:944: warning: missing braces around initializer
13:03:03 zephyr/subsys/logging/log_core.c:944: warning: (near initialization for ‘__init_enable_logger.
```

cc: @dcpleung , @andyross , @kv2019i

cfriedt replied:
@marc-hb - can you submit a PR? It's unlikely I will have a similar toolchain at my disposal to test.

marc-hb replied:
Absolutely; I'd love to test and submit a change but first I need to understand which "advanced" (> C99) struct initialization feature is at play here. Either I'm a C99 dinosaur or the "macrobatics" confuse me or most likely: a bit of both. Care to share a feature name / search keyword / C version this initialization style relies on? If nothing else, an example of how the code looks after macro expansion. Any clue appreciated.

fabiobaltieri replied:
`warning: missing braces around initializer` could it need an extra pair of braces around `init_fn_`? `COND_CODE_1` always needs extra ones, never understood why, wondering if if they get eaten up by the macro processor.

fabiobaltieri replied:
the expansion for me is exactly the same before and after btw

`__attribute__((__section__( ".z_init_" "POST_KERNEL" "90""_" "00019""_"))) __init___device_dts_ord_19 = { .init_fn = {.dev = (&led_gpio_init)}, .dev = &__device_dts_ord_19, };` (this is `build/zephyr/drivers/led/CMakeFiles/drivers__led.dir/led_gpio.c.i` with `COMPILER_SAVE_TEMPS=y`)

fabiobaltieri commented on include/zephyr/init.h:
```diff
@@ -96,7 +107,12 @@ struct init_entry {
 	 * reference to it, otherwise it is set to NULL.
 	 */
-	const struct device *dev;
+	union {
+		const struct device *dev;
+#ifdef CONFIG_DEVICE_MUTABLE
+		struct device *dev_rw;
+#endif
+	};
```
@marc-hb `__init_enable_logger` is a `SYS_INIT`, I think `unknown field ‘dev’ specified in initializer` is referring to this, and it may need the brace to find the field in the union.

fabiobaltieri replied:
was able to repro: https://godbolt.org/z/Whf44q63s does not work on GCC 4, I'll send a PR, easy enough

fabiobaltieri replied:
https://github.com/zephyrproject-rtos/zephyr/pull/65912
